server/src/main/java/io/druid/server/http/NamespacesResource.java

```diff
@@ -0,0 +1,242 @@
+  private final CuratorFramework curator;
+  private ZkPathsConfig zkPathsConfig;
+
+  @Inject
+  public NamespacesResource(
+      @Json ObjectMapper jsonMapper,
+      @Smile ObjectMapper smileMapper,
+      CuratorFramework curator,
+      ZkPathsConfig zkPathsConfig
+  )
+  {
+    this.jsonMapper = jsonMapper;
+    this.smileMapper = smileMapper;
+    this.curator = curator;
+    this.zkPathsConfig = zkPathsConfig;
+  }
+
+  @GET
+  @Produces({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE})
+  public Response getNamespaces()
+  {
+    Response.ResponseBuilder builder = Response.ok();
+    try {
+      builder.entity(
+          Collections2.filter(
+              Collections2.transform(
+                  curator.getChildren().forPath(zkPathsConfig.getNamespacePath()),
+                  new Function<String, Map<String, Object>>()
+                  {
+                    @Nullable
+                    @Override
+                    public Map<String, Object> apply(String input)
+                    {
+                      final String path = ZKPaths.makePath(zkPathsConfig.getNamespacePath(), input);
+                      try {
+                        final byte[] data = curator.getData().forPath(path);
+                        if (data == null) {
+                          // Race condition... missing path
+                          return null;
+                        }
+                        return jsonMapper.readValue(
+                            data, new TypeReference<Map<String, Object>>()
+                            {
+                            }
+                        );
+                      }
+                      catch (Exception e) {
+                        log.warn(e, "Error on path [%s]", path);
+                        return null;
+                      }
+                    }
+                  }
+              ), new Predicate<Map<String, Object>>()
+              {
+                @Override
+                public boolean apply(@Nullable Map<String, Object> input)
+                {
+                  return input != null;
+                }
+              }
+          )
+      );
+    }
+    catch (Exception e) {
+      log.error(e, "Error getting list of namespaces");
+      return Response.serverError().entity(cleanError(e)).build();
+    }
+    return builder.build();
+  }
+
+  @POST
+  @Produces({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE})
+  @Consumes({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE})
+  public Response newNamespace(
+      InputStream inputStream,
+      @Context final HttpServletRequest req // used only to get request content-type
+  )
+  {
+    final String reqContentType = req.getContentType();
+    final boolean isSmile = SmileMediaTypes.APPLICATION_JACKSON_SMILE.equals(reqContentType);
+    final ObjectMapper mapper = isSmile ? smileMapper : jsonMapper;
```

Comment:
It would be really nice if this stuff could move into some Jersey Serde thingie and this method could just depend on an actual Java object that we know can be deserialized on the historical/broker side and turned into something meaningful.